File: iosApp/iosApp/Component/AuctionRow.swift
import SwiftUI
import Shared

struct AuctionRow: View {
    let auction: Auction

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            // Image + Name + Button
            HStack(spacing: 8) {
                ProductThumbnail(url: auction.product.mainImage?.imageUrlThumb)

                VStack(alignment: .leading, spacing: 4) {
                    Text(auction.product.name)
                        .font(.body)
                        .fontWeight(.semibold)
                        .lineLimit(1)

                    Spacer()
                }
                .frame(height: 100)
            }

            // Description
            Text(auction.product.description_ ?? "")
                .font(.footnote)
                .lineLimit(2)

            // Municipality
            Text(auction.product.municipalityName)
                .font(.footnote)

            // Current bid
            Text("Current bid: \(auction.product.currentBid)")
                .font(.body)
                .fontWeight(.semibold)

            // Categories
            VStack(alignment: .leading, spacing: 4) {
                ForEach([auction.category1, auction.category2, auction.category3].compactMap { $0 }, id: \.id) { category in
                    Text(category.headline)
                        .foregroundColor(.blue)
                }
            }

            // End date
            Text(auction.product.endDate)
                .font(.footnote)
                .fontWeight(.semibold)
        }
        .padding(12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color(uiColor: .systemBackground))
        .clipShape(RoundedRectangle(cornerRadius: 8))
        .shadow(radius: 2)
    }
}

struct ProductThumbnail: View {
    var url: String?

    var body: some View {
        ZStack {
            Color.gray.opacity(0.2)
                .frame(width: 100, height: 100)
                .clipShape(RoundedRectangle(cornerRadius: 8))
            
            Group {
                if let urlString = url, let imageURL = URL(string: urlString) {
                    

                    AsyncImage(url: imageURL) { phase in
                        switch phase {
                        case .empty:
                            ProgressView()
                                .frame(width: 100, height: 100)
                        case .success(let image):
                            image
                                .resizable()
                                .scaledToFill()
                                .frame(width: 100, height: 100)
                                .clipped()
                        case .failure:
                            EmptyView()
                        @unknown default:
                            EmptyView()
                        }
                    }
                } else {
                    Image(systemName: "lock.fill")
                        .resizable()
                        .scaledToFit()
                        .frame(width: 100, height: 100)
                        .foregroundColor(.gray)
                }
            }
            .clipShape(RoundedRectangle(cornerRadius: 8))
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/jetbrains/kmpapp/component/AuctionRow.kt
package com.jetbrains.kmpapp.component

import androidx.compose.foundation.Image
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import coil3.compose.rememberAsyncImagePainter
import com.jetbrains.kmpapp.data.Auction

@Composable
fun AuctionRow(
    modifier: Modifier = Modifier,
    auction: Auction,
    onClick: (Int) -> Unit,
) {
    Card(
        modifier = modifier.clickable { onClick(auction.product.id) },
        elevation = CardDefaults.cardElevation(4.dp),
        shape = RoundedCornerShape(8.dp),
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(8.dp),
            verticalArrangement = Arrangement.spacedBy(4.dp),
            horizontalAlignment = Alignment.Start,
        ) {
            // image and the name
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                // thumbnail image
                ProductThumbnail(url = auction.product.mainImage?.imageUrlThumb)

                Box(
                    modifier = Modifier.height(100.dp).weight(1f)
                ) {

                    // name
                    Text(
                        modifier = Modifier.align(Alignment.TopStart),
                        text = auction.product.name,
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.SemiBold,
                        maxLines = 1,
                    )
                }
            }

            // description
            Text(
                text = auction.product.description ?: "",
                style = MaterialTheme.typography.bodySmall,
                maxLines = 2,
            )

            // municipality name
            Text(
                text = auction.product.municipalityName,
                style = MaterialTheme.typography.bodySmall,
                maxLines = 1,
            )

            // current bid
            Text(
                text = "Current bid: ${auction.product.currentBid}",
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.SemiBold,
                maxLines = 1,
            )

            // categories
            Column(
                verticalArrangement = Arrangement.spacedBy(4.dp),
                horizontalAlignment = Alignment.Start,
            ) {
                // category 1
                auction.category1?.let {
                    Text(
                        text = it.headline,
                        color = Color.Blue,
                    )
                }

                // category 2
                auction.category2?.let {
                    Text(
                        text = it.headline,
                        color = Color.Blue,
                    )
                }

                // category 3
                auction.category3?.let {
                    Text(
                        text = it.headline,
                        color = Color.Blue,
                    )
                }
            }

            // end date
            Text(
                text = auction.product.endDate,
                style = MaterialTheme.typography.bodySmall,
                fontWeight = FontWeight.SemiBold,
                maxLines = 1,
            )
        }
    }
}

@Composable
fun ProductThumbnail(url: String?) {
    if (url != null) {
        Image(
            modifier = Modifier.size(100.dp).clip(RoundedCornerShape(8.dp)),
            painter = rememberAsyncImagePainter(model = url),
            contentDescription = "product thumbnail",
            contentScale = ContentScale.Crop
        )
    } else {
        Image(
            imageVector = Icons.Default.Lock,
            contentDescription = "product placeholder",
            modifier = Modifier.size(100.dp).clip(RoundedCornerShape(8.dp)),
            contentScale = ContentScale.Crop
        )
    }
}